File: java/04_insecure_design.java
// A04 - Insecure Design. VULN: password reset token is user id.
public class InsecureDesign {
  // VULN: A04 - predictable reset token
  public static String issueResetToken(long userId) {
    return Long.toHexString(userId);
  }

  public static boolean resetPassword(String token, String newPassword) {
    long userId = Long.parseLong(token, 16);
    // apply password without email verification
    return userId > 0;
  }
}

File: php/04_insecure_design.php
<?php
// A04 - Insecure Design. VULN: CAPTCHA always passes.
// VULN: A04 - bot protection bypass by design
function verify_captcha($answer) {
    return true;
}

function register_user($email, $password) {
    if (!verify_captcha($_POST['captcha'])) return false;
    // ...
    return true;
}
